server/src/main/java/org/elasticsearch/common/geo/CentroidCalculator.java

```diff
@@ -135,17 +151,42 @@ public Void visit(GeometryCollection<?> collection) {
 
         @Override
         public Void visit(Line line) {
-            for (int i = 0; i < line.length(); i++) {
-                calculator.addCoordinate(line.getX(i), line.getY(i));
+            // a line's centroid is calculated by summing the center of each
+            // line segment weighted by the line segment's length in degrees
+            for (int i = 0; i < line.length() - 1; i++) {
+                double diffX = line.getX(i) - line.getX(i + 1);
+                double diffY = line.getY(i) - line.getY(i + 1);
+                double x = (line.getX(i) + line.getX(i + 1)) / 2;
+                double y = (line.getY(i) + line.getY(i + 1)) / 2;
+                calculator.addCoordinate(x, y, Math.sqrt(diffX * diffX + diffY * diffY));
             }
             return null;
         }
 
         @Override
         public Void visit(LinearRing ring) {
+            // implementation of calculation defined in
+            // https://www.seas.upenn.edu/~sys502/extra_materials/Polygon%20Area%20and%20Centroid.pdf
+            //
+            // centroid of a ring is a weighted coordinate based on the ring's area.
+            // sign of area is determined by orientation: positive for CCW, negative for CW.
+
+            double totalRingArea = 0.0;
+            for (int i = 0; i < ring.length() - 1; i++) {
```

Comment:
I am having doubts here as we accept polygons with both orientations so the sign area can be negative when it should be positive. In any case I think it should be only positive, we should add test with both orientations.

Reply:
ah we do! forgot about that while working on this. I originally had support for this, but reverted it. will re-add and add tests. just need to abs() the areas and keep track of holes vs. outer-ring